src/initio/initio.c

```diff
@@ -0,0 +1,794 @@
+      mraa_uart_stop(dev);
+      return NULL;
+    } else {
+      if (++idx == n) return dev;
+    }
+  } else {
+    syslog(LOG_ERR, "parse_uart: invalid uart baudrate");
+  }
+
+  /* Check for mode - [int] bytesize, [mraa_uart_parity_t] parity, [int] stopbits*/
+  char* end = proto[idx];
+  /* Check for bytesize*/
+  int bytesize = -1;
+  if (mraa_atoi_x(proto[idx], end, &bytesize, 0) != MRAA_SUCCESS) {
+    syslog(LOG_ERR, "parse_uart: error setting up uart bytesize %d", bytesize);
+    mraa_uart_stop(dev);
+    return NULL;
+  }
+
+  int parity = -1;
+  if (strncmp(end, U_PARITY_NONE, strlen(U_PARITY_NONE)) == 0) {
+    parity = MRAA_UART_PARITY_NONE;
+    end+=strlen(U_PARITY_NONE);
+  } else if (strncmp(end, U_PARITY_EVEN, strlen(U_PARITY_EVEN)) == 0) {
+    parity = MRAA_UART_PARITY_EVEN;
+    end+=strlen(U_PARITY_EVEN);
+  } else if (strncmp(end, U_PARITY_ODD, strlen(U_PARITY_ODD)) == 0) {
+    parity = MRAA_UART_PARITY_ODD;
+    end+=strlen(U_PARITY_ODD);
+  } else if (strncmp(end, U_PARITY_MARK, strlen(U_PARITY_MARK)) == 0) {
+    parity = MRAA_UART_PARITY_MARK;
+    end+=strlen(U_PARITY_MARK);
+  } else if (strncmp(end, U_PARITY_SPACE, strlen(U_PARITY_SPACE)) == 0) {
+    parity = MRAA_UART_PARITY_SPACE;
+    end+=strlen(U_PARITY_SPACE);
+  } else {}
+
+  if (parity == -1) {
+    syslog(LOG_ERR, "parse_uart: error setting up uart parity");
+    mraa_uart_stop(dev);
+    return NULL;
+  }
+
+  int stopbits = -1;
+  if (mraa_atoi_x(end, NULL, &stopbits, 0) != MRAA_SUCCESS) {
+    syslog(LOG_ERR, "parse_uart: error setting up uart bytesize %d", bytesize);
+    mraa_uart_stop(dev);
+    return NULL;
+  }
+
+  if (mraa_uart_set_mode(dev, bytesize, (mraa_uart_parity_t)parity, stopbits) != MRAA_SUCCESS) {
+    syslog(LOG_ERR, "parse_uart: error setting up uart mode");
+    mraa_uart_stop(dev);
+    return NULL;
+  }
+
+  return dev;
+}
+
+static mraa_spi_context parse_spi(char **proto, size_t n)
+{
+  if (proto == NULL || n <= 1) {
+    syslog(LOG_ERR, "parse_spi: invalid parameters");
+    return NULL;
+  }
+
+  int idx = 1;
+  mraa_spi_context dev = NULL;
+
+  int bus = -1;
+  if (proto[idx] && (mraa_atoi_x(proto[idx], NULL, &bus, 0) == MRAA_SUCCESS)) {
+    dev = mraa_spi_init(bus);
+    if (dev == NULL) {
+      syslog(LOG_ERR, "parse_spi: could not init spi bus %d", bus);
+      return NULL;
+    } else {
+      if (++idx == n) return dev;
+    }
+  } else {
+    syslog(LOG_ERR, "parse_spi: invalid spi bus number");
+    return NULL;
+  }
+
+  /* Check for mode. */
+  int mode = -1;
+  if (strncmp(proto[idx], S_MODE_0, strlen(S_MODE_0)) == 0) {
+    mode = MRAA_SPI_MODE0;
+  } else if (strncmp(proto[idx], S_MODE_1, strlen(S_MODE_1)) == 0) {
+    mode = MRAA_SPI_MODE1;
+  } else if (strncmp(proto[idx], S_MODE_2, strlen(S_MODE_2)) == 0) {
+    mode = MRAA_SPI_MODE2;
+  } else if (strncmp(proto[idx], S_MODE_3, strlen(S_MODE_3)) == 0) {
+    mode = MRAA_SPI_MODE3;
+  } else {}
```

Comment:
Why the empty clause? Please remove if really unused. Same goes for all repeats below (I won't comment on all of them)

Reply:
Some Misra C rules dictate that. Will remove them

Reply:
Yeah, I suspected that :) Makes no sense in mraa context, so yes, please.